### ISSUE TITLE
Changing a users password produces a nullreference error

### ISSUE THREAD

**Issue by srothery** [FusionAuth/fusionauth-issues#758]
### Description 
When the edit user UI is used to change a user's password and the save button is pressed a message pops up saying `FusionAuth encountered an unexpected error. Please contact support for assistance`. Checking the log shows an invalidReference exception and a long series of errors reported (see attached [fusionauthlog.txt](https://github.com/FusionAuth/fusionauth-issues/files/4947234/fusionauthlog.txt)). Repeating with using postman to talk to the API direct gives the same issue. This is using the docker image from https://hub.docker.com/r/fusionauth/fusionauth-app. I have tested this with predefined users and new users created via the API and both can't have their passwords changed using either the admin UI or the API.

### Affects versions
I have reproduced this with a clean container muliple times using version 1.18.1 and 1.17.5.
Version 1.16.1 does not have this error (or any warnings) and this is the version we are going with for now.
No other versions tested.

### Steps to reproduce
1. Create fresh docker image with a two users predefined using the bootstrap json
2. Login to the fusionauth UI using the admin
3. Navigate to the normal user
4. Click edit user
5. Select the drop down to change the password
6. type and retype the new password
7. hit the save button

### Expected behavior
This appears to be a regression from the 1.16.1 behaviour. I'd expect the password to be changed without error.

### Platform
(Please complete the following information)
 - Device: Laptop
 - OS: Centos 7 VM, docker 19.03.11, docker-compose 1.25.0
 - Browser + version : Firefox 78.0.2
 - Database:  Postgres 13

### Additional context
First found by my colleague Harry: https://fusionauth.io/community/forum/topic/225/all-password-change-apis-failing

**Comment by robotdan** [comment]
> Create fresh docker image with a two users predefined using the bootstrap json


Do you mean you are using Kickstart? If so, can you post your kickstart JSON file with any sensitive information removed? 

The only way this exception could occur is if the tenant does not have a refresh token usage policy. However, in the 1.17.0 migration this value should be set, and it is also defaulted during a create or update. 

> I have reproduced this with a clean container muliple times using version 1.18.1 and 1.17.5.
Version 1.16.1 does not have this error (or any warnings) and this is the version we are going with for now.

When you are testing, are you using the default tenant, or creating a separate tenant? If so, does it occur in both the default tenant and the new tenant?

**Comment by srothery** [comment]
> Do you mean you are using Kickstart? If so, can you post your kickstart JSON file with any sensitive information removed?

Yes - I've attached my kickstart file with keys redacted.
 [kickstart.txt](https://github.com/FusionAuth/fusionauth-issues/files/4986596/kickstart.txt)

> When you are testing, are you using the default tenant, or creating a separate tenant? If so, does it occur in both the default tenant and the new tenant?

We are just using the default tenant.

**Comment by srothery** [comment]
We are still running 1.16.1 but will try later in case 1.18.7 doesn't show this issue

**Comment by mdwagner** [comment]
@robotdan I can confirm this issue only happens with the default tenant for me (fusionauth version: 1.19.8).

**Comment by ruckc** [comment]
Running into this also.  Kickstarted user.  1.20.0, kubernetes, default tenant.

Log output
```
2020-10-27T16:50:20.328535762Z 2020-10-27 4:50:20.326 PM ERROR io.fusionauth.app.primeframework.error.ExceptionExceptionHandler - An unhandled exception was thrown
2020-10-27T16:50:20.328577953Z java.lang.NullPointerException: null
2020-10-27T16:50:20.328583962Z 	at io.fusionauth.api.service.user.DefaultUserService.deleteTokensAndIdsOnPasswordChange(DefaultUserService.java:2178)
2020-10-27T16:50:20.328588823Z 	at io.fusionauth.api.service.user.DefaultUserService.updateAllowConnectorIdChange(DefaultUserService.java:1125)
2020-10-27T16:50:20.328593333Z 	at org.mybatis.guice.transactional.TransactionalMethodInterceptor.invoke(TransactionalMethodInterceptor.java:96)
2020-10-27T16:50:20.328597443Z 	at org.mybatis.guice.transactional.TransactionalMethodInterceptor.invoke(TransactionalMethodInterceptor.java:96)
2020-10-27T16:50:20.328602603Z 	at org.mybatis.guice.transactional.TransactionalMethodInterceptor.invoke(TransactionalMethodInterceptor.java:96)
2020-10-27T16:50:20.328606573Z 	at org.mybatis.guice.transactional.TransactionalMethodInterceptor.invoke(TransactionalMethodInterceptor.java:96)
2020-10-27T16:50:20.328610613Z 	at io.fusionauth.api.service.user.DefaultUserService.update(DefaultUserService.java:1103)
2020-10-27T16:50:20.328614493Z 	at org.mybatis.guice.transactional.TransactionalMethodInterceptor.invoke(TransactionalMethodInterceptor.java:96)
2020-10-27T16:50:20.328618723Z 	at org.mybatis.guice.transactional.TransactionalMethodInterceptor.invoke(TransactionalMethodInterceptor.java:96)
2020-10-27T16:50:20.328622593Z 	at org.mybatis.guice.transactional.TransactionalMethodInterceptor.invoke(TransactionalMethodInterceptor.java:96)
2020-10-27T16:50:20.328626613Z 	at org.mybatis.guice.transactional.TransactionalMethodInterceptor.invoke(TransactionalMethodInterceptor.java:96)
2020-10-27T16:50:20.328630624Z 	at io.fusionauth.app.action.api.UserAction.put(UserAction.java:155)
2020-10-27T16:50:20.328634323Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
```

Kickstart example:
```json
          {
            "method": "POST",
            "url": "/api/user/registration",
            "body": {
              "user": {
                "email": "#{user01Email}",
                "username": "#{user01username}",
                "firstName": "#{user01firstName}",
                "middleName": "#{user01middleName}",
                "lastName": "#{user01lastName}",
                "fullName": "#{user01fullName}",
                "password": "#{defaultPassword}"
              },
              "registration": {
                "applicationId": "#{FUSIONAUTH_APPLICATION_ID}",
                "roles": [
                  "admin"
                ]
              }
            }
          },
```

**Comment by srothery** [comment]
@ruckc if you create your own tenant rather than use the default we have found it avoids this issue.

**Comment by tumauwt** [comment]
Verified that setting a user to "Require Password Change" also produces the same error.

**Comment by ruckc** [comment]
Looking into this further, yes it's the default tenant, and FA admin's can't change their password, and FA admins have to exist in the default tenant, based on our testing.

**Comment by tumauwt** [comment]
The above testing referenced by ruckc was conducted on 1.20.0. 

From what we have been able to test, you need at least one admin user in the FusionAuth App in order to create additional admins via the GUI. We were unable to create the FusionAuth App in a separate tenant. 

Even tried setting defaultTenantId to override the generated value to see if that would change the behavior - it did not.

**Comment by robotdan** [comment]
Haven't recreated yet, but the stack grace from @ruckc looks like the Tenant is in a strange state. To work around, try touching the Tenant in the UI, Edit, Save. 

@ruckc is that you're entire kickstart file that you used to recreate the issue?

**Comment by mooreds** [comment]
Hey folks,

I was able to recreate this. 

* Used the attached kickstart file, fresh docker install
* version 1.20
* After you sign in as the admin user, go directly to the user screen and try to change your own password
* See the error message.

Stack trace:

```
fusionauth_1  | 2020-10-28 9:56:24.579 PM ERROR io.fusionauth.app.primeframework.error.ExceptionExceptionHandler - An unhandled exception was thrown
fusionauth_1  | java.lang.NullPointerException: null
fusionauth_1  | 	at io.fusionauth.api.service.user.DefaultUserService.deleteTokensAndIdsOnPasswordChange(DefaultUserService.java:2178)
fusionauth_1  | 	at io.fusionauth.api.service.user.DefaultUserService.updateAllowConnectorIdChange(DefaultUserService.java:1125)
fusionauth_1  | 	at org.mybatis.guice.transactional.TransactionalMethodInterceptor.invoke(TransactionalMethodInterceptor.java:96)
fusionauth_1  | 	at org.mybatis.guice.transactional.TransactionalMethodInterceptor.invoke(TransactionalMethodInterceptor.java:96)
fusionauth_1  | 	at org.mybatis.guice.transactional.TransactionalMethodInterceptor.invoke(TransactionalMethodInterceptor.java:96)
fusionauth_1  | 	at org.mybatis.guice.transactional.TransactionalMethodInterceptor.invoke(TransactionalMethodInterceptor.java:96)
fusionauth_1  | 	at io.fusionauth.api.service.user.DefaultUserService.update(DefaultUserService.java:1103)
```

The workaround is to go to the default tenant in the UI and save it (you don't even have to make any changes, just edit the tenant and save it). 

You can even PATCH the tenant via the API without changing a thing and that ensures the error won't occur:

```
curl -XPATCH -H 'Content-type: application/json' -H "Authorization: $API_KEY" 'http://localhost:9011/api/tenant/30663132-6464-6665-3032-32646661393' -d '{}'
```

Looks like there is some tenant setting which is not set by kickstart. Note that the kickstart file is dos formatted (not sure if that is an issue).

**Comment by robotdan** [comment]
@mooreds 

> As a work around please try to reproduce w/out changing the tenantId in the kickstart file.



Does that change the behavior?

**Comment by mooreds** [comment]
@robotdan unfortunately, no. I still see the error if I go directly to the user details page and try to change the password:

```
fusionauth_1  | 2020-10-28 10:52:20.403 PM ERROR io.fusionauth.app.primeframework.error.ExceptionExceptionHandler - An unhandled exception was thrown
fusionauth_1  | java.lang.NullPointerException: null
fusionauth_1  | 	at io.fusionauth.api.service.user.DefaultUserService.deleteTokensAndIdsOnPasswordChange(DefaultUserService.java:2178)
fusionauth_1  | 	at io.fusionauth.api.service.user.DefaultUserService.updateAllowConnectorIdChange(DefaultUserService.java:1125)
fusionauth_1  | 	at org.mybatis.guice.transactional.TransactionalMethodInterceptor.invoke(TransactionalMethodInterceptor.java:96)
fusionauth_1  | 	at org.mybatis.guice.transactional.TransactionalMethodInterceptor.invoke(TransactionalMethodInterceptor.java:96)
```

I also see this error:

```
fusionauth_1  | 2020-10-28 10:54:11.457 PM WARN  org.primeframework.mvc.action.result.SavedRequestTools - Bad SavedRequest cookie [bhZazWJxwNEex2wLHjs4eaF8ZeTpCtUYgaFHUo7rvdu41SlRVETbzR1H_nR6OT39c8EL7mQx1Z_7rk8TB8YNZ1Se6VPaqswpLwDoXLoY3hNR7g8kVubASYrqUUQvyBoVcGFEIcop9kIqCHwPq0kun1iQfTswMH2NSf6knqrpJGE=]. Error is [Given final block not properly padded. Such issues can arise if a bad key is used during decryption.]
fusionauth_1  | 2020-10-28 10:54:11.711 PM WARN  org.primeframework.mvc.action.result.SavedRequestTools - Bad SavedRequest cookie [bhZazWJxwNEex2wLHjs4eaF8ZeTpCtUYgaFHUo7rvdu41SlRVETbzR1H_nR6OT39c8EL7mQx1Z_7rk8TB8YNZ1Se6VPaqswpLwDoXLoY3hNR7g8kVubASYrqUUQvyBoVcGFEIcop9kIqCHwPq0kun1iQfTswMH2NSf6knqrpJGE=]. Error is [Given final block not properly padded. Such issues can arise if a bad key is used during decryption.]
```

I tried using an incognito firefox window and am now seeing this error message when I edit the user (in the dev console):

```
 0 LastLogins.js:61:13
    _handleAjaxError LastLogins.js:61
    _handler PrimeRequest.js:485
    go PrimeRequest.js:85
    update LastLogins.js:30
    LastLogins LastLogins.js:22
    ManageUser ManageUser.js:20
    <anonymous> 00000000-0000-0000-0000-000000000001:122
    _callReadyListeners PrimeDocument.js:377
    (Async: EventListener.handleEvent)
    onReady PrimeDocument.js:193
    Prime Overlay.js:110
    <anonymous> prime-es6-1.5.3.js:10306
```

And with this scenario (where I use the default tenant id and don't configure that in kickstart), my workaround of saving the tenant doesn't seem to resolve the issue.

**Comment by robotdan** [comment]
The `org.primeframework.mvc.action.result.SavedRequestTools` error is just noise, you have a save request cookie from a previous instance of FusionAuth that used a different encryption key. 

Still unable to recreate in a test, even with using that kickstart JSON.

**Comment by mooreds** [comment]
Hmmm. Here's my exact files if that helps.

docker compose:
```
```

**Comment by mooreds** [comment]
I looked into the postgresql database and noticed the following on the tenant configuration:

```
# when login fails
"jwtConfiguration": {"enabled": true, "timeToLiveInSeconds": 3600, "refreshTokenExpirationPolicy": "Fixed", "refreshTokenTimeToLiveInMinutes": 43200, "refreshTokenUsagePolicy": "Reusable"},

# after I edit the tenant one time, not changing anything, password change succeeds

"jwtConfiguration":{"enabled":true,"refreshTokenExpirationPolicy":"Fixed","refreshTokenRevocationPolicy":{"onLoginPrevented":true,"onPasswordChanged":true},"refreshTokenTimeToLiveInMinutes":43200,"refreshTokenUsagePolicy":"Reusable","timeToLiveInSeconds":3600},
```

**Comment by robotdan** [comment]
Thanks @mooreds found it! Your comment made me look at something I already looked and had convinced myself it wasn't the problem.... but it was.

**Comment by tumauwt** [comment]
Another work around is:
Add a PATCH request to the default tenant in your kickstart file after api creation. Users are then able to change passwords without having to "save" the default tenant. Verified - eg below
          {
            "method": "PATCH",
            "url": "/api/tenant/#{defaultTenantId}",
            "body": {
              "tenant": {
                "name": "Some New Name"
              }
            }
          },

**Comment by tumauwt** [comment]
Guess the work around is now OBE.

**Comment by robotdan** [comment]
Thanks for the suggestion @tumauwt - I was just about to add this comment as well to document that a `PATCH` request in your kickstart configuration would also fix the problem. 

You beat me to it! Thanks!

**Comment by srothery** [comment]
Good to see a number of workarounds that still allow the default tenant. With this closed does that mean no fix ?

**Comment by robotdan** [comment]
@srothery this was fixed in 1.20.1.
